adapters/http-vertx-base/src/main/java/org/eclipse/hono/adapter/http/AbstractVertxBasedHttpProtocolAdapter.java

```diff
@@ -315,16 +316,24 @@ protected Router createRouter() {
         final TracingHandler tracingHandler = createTracingHandler();
         matchAllRoute.handler(tracingHandler).failureHandler(tracingHandler);
 
-        // 3. default handler for failed routes
+        // 3. handler to ensure tracing root span is finished if the request gets closed prematurely
+        matchAllRoute.handler(ctx -> {
+            if (!ctx.response().closed() && !ctx.response().ended()) {
+                ctx.response().closeHandler(v -> finishRootSpanOnResponseGettingClosedPrematurely(ctx));
```

Comment:
the close handler that is set here will be overwritten later by the *doUploadMessage* method, right?
Why don't we set a failure Handler here instead which finishes the root span and modify the *doUploadMessage* method to fail the RoutingContext e.g. with an IllegalStateException if the request is being closed before we can send the response or if sending the response fails?

Reply:
> the close handler that is set here will be overwritten later by the doUploadMessage method, right?

Yes, that's intended.

> Why don't we set a failure Handler here instead which finishes the root span and modify the doUploadMessage method to fail the RoutingContext e.g. with an IllegalStateException if the request is being closed before we can send the response or if sending the response fails?

Hmm, I'll think about that. I'm currently doing some more testing.